### ISSUE TITLE
Unable to use transactions with multiple DbContexts for the same database

### ISSUE THREAD

**Issue by sabvente** [PomeloFoundation/Pomelo.EntityFrameworkCore.MySql#1040]
Hello,
I'm trying to use a transaction across multiple DbContexts with the same connection string.

I'm getting the following error after calling the second context:
`The specified transaction is not associated with the current connection. Only transactions associated with the current connection may be used.`

According to `Microsoft.EntityFrameworkCore.Database.Connection:Debug` logs, there is actually a second db connection opened when I call the second context.

Pseuodo-code:
```csharp
using (var currentTransaction = contextA.Database.BeginTransaction())
{
    // query contextA
    contextB.Database.UseTransaction(currentTransaction.GetDbTransaction()); // Exception
    // query contextB
    currentTransaction.Commit();
}
```

Each DbContext is added like this:
```csharp
services.AddDbContext<MyDbContext>(dbContextOptionsBuilder =>
{
    dbContextOptionsBuilder.UseMySql(configuration.GetConnectionString("Database"));
});
```
This is an ASP.Net Core 2.2 project with basic configuration.
Pomelo.EntityFrameworkCore.MySql version: `2.2.6`

According to my understanding, the connection opened for the first context should be also used for the second context, because they have the same connection string so the connection can be reused (in the scope of the same request). If that is wrong, how can I reuse the connection?

**Comment by lauxjpn** [comment]
> the connection opened for the first context should be also used for the second context, because they have the same connection string so the connection can be reused

This seems like a logical conclusion at first, but it is not the case. If you do explicitly supply the same connection to two separate context instances, than each of them will use its own connection instance, even when both use an identical connection string.

Connections might be reused, if connection pooling is enabled, but only after the connection instance got closed by the user.

> how can I reuse the connection?

You should find all you need to know in the [Using Transactions: Cross-context transaction (relational databases only)](https://docs.microsoft.com/en-us/ef/core/saving/transactions#cross-context-transaction-relational-databases-only) article of the official EF Core docs. Of course you will need to replace all `Sql` prefixes with `MySql` in their sample code.

Basically, you need to instantiate a `MySqlConnection` first and use that same object for each of the `UseMySql` calls.

**Comment by sabvente** [comment]
Hi @lauxjpn,

Thank you for your answer.
> each of them will use its own connection instance, even if both are created from an identical connection string.

That explains why I'm getting the specific exception.

>Basically, you need to instantiate a MySqlConnection first (using your connection string) and use that same connection object for each of the UseMySql calls (instead of a connection string).

Oh, I really hoped that I was missing some inner knowledge of EF/EFMySql. It just don't feel "right" to create `MySqlConnection` manually and handle Open/Close of the connection.
The ef core docs don't provide sample for manually creating connection in an ASP.Net DI environment.

I will move forward and create a new service to manage the dbconnection.

**Comment by lauxjpn** [comment]
> The ef core docs don't provide sample for manually creating connection in an ASP.Net DI environment.

You can use something like the following to let the DI container handle the connection:

```c#
public void ConfigureServices(IServiceCollection services)
{
    services.AddScoped<DbConnection>(sp => new MySqlConnection(
                    Configuration.GetConnectionString("MyConnectionString")));
}
```

You can then access the connection as usual via constructor injection (or by manually calling `GetService` if you must).
